aten/src/ATen/native/native_functions.yaml

```diff
@@ -13819,13 +13819,21 @@
   variants: function
   autogen: _scaled_dot_product_attention.out
 
+- func: _scale_factor_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, Scalar scale_factor=1.0) -> Tensor
+  python_module: nn
+  variants: function
+  autogen: _scale_factor_dot_product_attention.out
+
 # This aten function is kept so that we can test the choice function from Python
 - func: _fused_sdp_choice(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False) -> int
   dispatch:
     Meta: _fused_sdp_choice_meta
     CPU, NestedTensorCPU: _fused_sdp_choice_cpp
     CUDA, NestedTensorCUDA: _fused_sdp_choice_cuda
 
+- func: _scale_factor_dot_product_attention_math(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, Tensor? dropout_mask=None, Scalar scale_factor=1.0) -> (Tensor, Tensor)
```

Comment:
Adding the optional scale factor Kwarg to the original is something I have planned for the 2.1 release. We don't want to pile in more private functions if we don't have to.

Can you still handle a more limited set of patterns without the generic scale factor?

Reply:
> Can you still handle a more limited set of patterns without the generic scale factor?

No, being able to determine the scale factor at execution time is essential as it is applied in every version of scaled_dot_product. 

I can understand that it's kind of ugly to add that private function. I didn't want to mess with an existing public API, otherwise I would not have done that.

But, if you have planned to add the scale_factor kwarg anyway, I can easily modify and simplify this diff so that it uses a configurable function with the same signature as _scale_factor_dot_product_attention. Then the sdp rewrite functionality could be activated later once that function exists. How about that?

Reply:
Just for my understanding. User code may look like 
``` Python
q * k.T/(q.size(-1) + the rest of sdpa
``` 
You are saying that at FX graph rewrite time though, this will have decomposed into k.transpose().div(<Something>). You can't tell at rewrite time that this scale factor is indeed the one that is baked into SDPA?

Again this will be a null point once the update is landed but trying to see if a smaller unit of work is landable for right after the 2.0 branch cut.

Reply:
Yes, this has to do with two aspects. 

First of all, at the point of rewrite, I do not have information about the shapes of intermediate tensor results in the fx.Graph. Therefore, the specific value of q.size(-1) is unknown. Symbolically I could, of course, look for .div(sqrt(query.size(-1))), but this would not work reliably for two reasons: First of all, a user could just write .div(sqrt(EMBED_DIM)) where EMBED_DIM is some constant or computed in whatever way imaginable. 

Second, while torch.fx.symbolic_trace converts query.size(-1) to query.size(-1), torch dynamo seems to replace it by query.size(3) (based on shape inference on concrete inputs ). Only it doesn't store that info in the fx graph as far as I could determine ( still a noob on that codebase of course, so maybe I didn't find it, but there is a place for it in (IIRC) node.meta of the fx.Graph, and this was not populated with the neccessary info when I stepped through with a debugger in the corresponding functions in _inductor/overrides.py.

I really need to be able to ensure that q.size(-1) and whatever is divided by are equal. This is easiest at runtime and doesn't really produce any measurable overhead ( e.g. just a few CPU cycles). 

What might be possible is to use the @torch.fx.wrap mechanism to push the decision into python-land. What could happen then are issues popping up from torch dynamo trying to introspect, break up and optimize that function further ( leading to a possible endless recursion of optimization expansions as the replacement pattern would usually also match the replacements fallback implementation ).

Reply:
I just made the changes I proposed. The PR is way simpler now.